### PR TITLE
[WIP] NCCHContainer: support for partitions if container is NCSD

### PR DESCRIPTION
**Pull request by B3n30**
This allows games to open other partitions than partition 0 from NCSDs (aka .3ds files).

For that the NCSD header is parsed instead of being skipped. To open another partition we need to keep track of which (3ds) file is currently open on which process. On real HW the loader does something similar by registering the process and media type with FS:Reg::Register. 

We can skip the FS:Reg module and just store the information in FS_User, since there we need it.

I also added the FS_USER::GetSpecialContentIndex changes in here since they rely on the NCSD changes. So this superseds #5329

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5345)

### CODE REVIEW

File: src/core/hle/service/fs/archive.cpp [hunk 1]
```diff
@@ -31,6 +31,16 @@
 
 namespace Service::FS {
 
+MediaType GetMediaTypeFromPath(const std::string& path) {
```
Comment (lioncash) [inline src/core/hle/service/fs/archive.cpp:34]:
```suggestion
MediaType GetMediaTypeFromPath(std::string_view path) {
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 2]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
     rb.Push<u64>(0);      // the secure value
 }
 
+void FS_USER::Register(u32 process_id, u64 program_id, const std::string& filepath) {
+    MediaType media_type = GetMediaTypeFromPath(filepath);
+    program_info_map[process_id] = ProgramInfo{program_id, media_type};
```
Comment (lioncash) [inline src/core/hle/service/fs/fs_user.cpp:834]:
```suggestion
    program_info_map.insert_or_assign(process_id, ProgramInfo{program_id, media_type});
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 3]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
     rb.Push<u64>(0);      // the secure value
 }
 
+void FS_USER::Register(u32 process_id, u64 program_id, const std::string& filepath) {
+    MediaType media_type = GetMediaTypeFromPath(filepath);
```
Comment (lioncash) [inline src/core/hle/service/fs/fs_user.cpp:833]:
```suggestion
    const MediaType media_type = GetMediaTypeFromPath(filepath);
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 4]
```diff
@@ -752,6 +747,32 @@ void FS_USER::ObsoletedDeleteExtSaveData(Kernel::HLERequestContext& ctx) {
...
+    const auto type = rp.PopEnum<SpecialContentType>();
+
+    LOG_DEBUG(Service_FS, "called, media_type={:08X} type={:08X}, title_id={:016X}",
+              static_cast<u32>(media_type), static_cast<u32>(type), title_id);
+
+    ResultVal<u16> index;
+    if (media_type == MediaType::GameCard) {
+        index = GetSpecialContentIndexFromGameCard(title_id, type);
+    } else {
+        index = GetSpecialContentIndexFromTDM(media_type, title_id, type);
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.cpp:763]:
```suggestion
        index = GetSpecialContentIndexFromTMD(media_type, title_id, type);
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 5]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
...
+    case SpecialContentType::Manual:
+        return MakeResult(static_cast<u16>(NCSDContentIndex::Manual));
+    case SpecialContentType::DLPChild:
+        return MakeResult(static_cast<u16>(NCSDContentIndex::DLP));
+    default:
+        ASSERT(false);
+    }
+}
+
+ResultVal<u16> FS_USER::GetSpecialContentIndexFromTDM(MediaType media_type, u64 title_id,
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.cpp:865]:
```suggestion
ResultVal<u16> FS_USER::GetSpecialContentIndexFromTMD(MediaType media_type, u64 title_id,
```

File: src/core/hle/service/fs/fs_user.h [hunk 6]
```diff
@@ -577,6 +598,18 @@ class FS_USER final : public ServiceFramework<FS_USER, ClientSlot> {
      */
     void GetSaveDataSecureValue(Kernel::HLERequestContext& ctx);
 
+    static ResultVal<u16> GetSpecialContentIndexFromGameCard(u64 title_id, SpecialContentType type);
+    static ResultVal<u16> GetSpecialContentIndexFromTDM(MediaType media_type, u64 title_id,
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.h:602]:
```suggestion
    static ResultVal<u16> GetSpecialContentIndexFromTMD(MediaType media_type, u64 title_id,
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 7]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
     rb.Push<u64>(0);      // the secure value
 }
 
+void FS_USER::Register(u32 process_id, u64 program_id, const std::string& filepath) {
+    const MediaType media_type = GetMediaTypeFromPath(filepath);
+    program_info_map.insert_or_assign(process_id, ProgramInfo{program_id, media_type};
+    if (media_type == MediaType::GameCard) {
+        current_gamecard_path = filepath;
+    }ƒ
```
Comment (lioncash) [inline src/core/hle/service/fs/fs_user.cpp:837]: ƒ seems to have worked its way in here :p

```suggestion
    }
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 8]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
     rb.Push<u64>(0);      // the secure value
 }
 
+void FS_USER::Register(u32 process_id, u64 program_id, const std::string& filepath) {
+    const MediaType media_type = GetMediaTypeFromPath(filepath);
+    program_info_map.insert_or_assign(process_id, ProgramInfo{program_id, media_type};
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.cpp:834]:
```suggestion
    program_info_map.insert_or_assign(process_id, ProgramInfo{program_id, media_type});
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 9]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
...
+
+    std::string tmd_path = AM::GetTitleMetadataPath(media_type, title_id);
+
+    FileSys::TitleMetadata tmd;
+    if (tmd.Load(tmd_path) != Loader::ResultStatus::Success || type == SpecialContentType::Update) {
+        // TODO(B3N30): Find correct result code
+        return ResultCode(-1);
+    }
+
+    // TODO(B3N30): Does real 3ds check if content exists in tmd?
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.cpp:880]:
```suggestion
    // TODO(B3N30): Does real 3DS check if content exists in TMD?
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 10]
```diff
@@ -808,6 +829,66 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
...
+
+std::string FS_USER::GetCurrentGamecardPath() const {
+    return current_gamecard_path;
+}
+
+ResultVal<u16> FS_USER::GetSpecialContentIndexFromGameCard(u64 title_id, SpecialContentType type) {
+    // TODO(B3N30) check if on real 3DS NCSD is checked if partition exists
+
+    if (type > SpecialContentType::DLPChild) {
+        // maybe type 4 is n3ds update/partition 6 but this needs more research
```
Comment (vvanelslande) [inline src/core/hle/service/fs/fs_user.cpp:848]:
```suggestion
        // Maybe type 4 is New 3DS update/partition 6 but this needs more research
```

**Comment by BreadFish64** [comment]
I just went ahead and fixed a conflict cause by merging #5397

**Pull request merged**